app/controllers/application_controller.rb

```diff
@@ -3,8 +3,161 @@
 class ApplicationController < Sinatra::Base
 
   configure do
+  	enable :sessions 
+  	set :session_secret, "jbsalt"
     set :public_folder, 'public'
     set :views, 'app/views'
   end
 
-end
\ No newline at end of file
+  get '/' do 
+  	erb :'index'
+  end 
+
+  get '/signup' do 
+  	if logged_in?
+  		redirect to '/tweets'
+  	else
+  		erb :'users/create_user' 
+  	end 
+  end 
+
+  post '/signup' do 
+  	# binding.pry
+  	if !params[:username].empty? && !params[:email].empty? && !params[:password].empty?
+	  	@user = User.create(username: params[:username], email: params[:email], password: params[:password])
+	  	session[:id] = @user.id 
+	  	redirect to "/tweets"
+	  else 
+	  	redirect to '/signup'
+	  end 
+  end 
+
+  get '/login' do 
+  	if !logged_in?
+  		erb :'users/login' 
+  	else
+  		redirect to '/tweets'
+  	end 
+  end 
+
+  post '/login' do 
+  	# binding.pry
+  	if logged_in? 
+  		redirect to '/tweets'
+  	else 
+	  	@user = User.find_by(username: params[:username])
+	  	if @user && @user.authenticate(params[:password])
+	  		session[:id] = @user.id 
+	  		redirect to "/tweets"
+	  	else 
+	  		redirect to '/login'
+	  	end 
+	  end 
+  end 
+
+  get '/users/:slug' do 
+  	@user = User.find_by_slug(params[:slug])
+  	erb :'/users/show'
+  end 
+
+  get '/logout' do 
+  	if logged_in?
+  		session.clear
+  	end 
+  	redirect to '/login' 
+  end 
+
+  get '/tweets' do 
+  	if logged_in?
+  		@user = User.find(session[:id])
+  		erb :'/tweets/tweets' 
+  	else
+  		redirect to '/login'
+  	end 
+  end 
+
+
+  get '/tweets/new' do 
+  	# binding.pry
+  	if logged_in?
+  		@user = User.find(session[:id])
+  		erb :'/tweets/create_tweet' 
+  	else
+  		redirect to '/login'
+  	end 
+  end 
+
+  post '/tweets' do 
+  	# binding.pry
+  	if !params[:content].empty?
+	  	@tweet = Tweet.create(content: params[:content])
+	  	@tweet.user = current_user
+	  	@tweet.save
+	  	current_user.save 
+	  	redirect to "/tweets/#{@tweet.id}"
+	  else 
+	  	redirect to '/tweets/new'
+	  end 
+  end 
+
+  get '/tweets/:id' do 
+  	if logged_in?
+	  	@tweet = Tweet.find(params[:id])
+	  	erb :'/tweets/show_tweet'
+	  else 
+	  	redirect to '/login'
+	  end 
+  end 
+
+  get '/tweets/:id/edit' do 
+  	if logged_in?
+  		@tweet = Tweet.find(params[:id])
+  		@user = @tweet.user  
+  		if session[:id] == @user.id
+  			erb :'/tweets/edit_tweet'
+  		else 
+  			redirect to '/tweets'
+  		end  
+  	else 
+  		redirect to '/login'
+  	end 
+  end 
+
+  post '/tweets/:id' do 
+  	if params[:content].empty?
+  		redirect to "/tweets/#{params[:id]}/edit"
+	  else 
+	  	@tweet = Tweet.find(params[:id])
+	  	@tweet.update(content: params[:content])
+	  	redirect to "/tweets/#{@tweet.id}"
+	  end 
+  end 
+
+  post '/tweets/:id/delete' do 
+  	if !logged_in? 
+  		redirect to '/login'
+  	end 
+  	@tweet = Tweet.find(params[:id]) 
+  	@user = @tweet.user  
+  	if session[:id] == @user.id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.